### PR TITLE
update pavement.py from upstream

### LINKED ISSUE #352: Travis is not working from upstream GeoServer
Relevant build:

https://travis-ci.org/kartoza/geonode/jobs/281121577

This is probably caused by upstream changes in GeoServer version.

Please investigate this first, there are several alternatives to resolve:

- Pin GeoServer version to specific version for 2.7.x
- Create changes for pavement.py handle the errors by inspecting upstream version of the file.

Comment on this ticket as the investigation goes.

We should fix this soon first.
@boney-bun can you investigate this?
I think it should be quick and only caused by packaged geoserver.

### PR DESCRIPTION
**Pull request by boney-bun**
fix #352

### CODE REVIEW

**Comment by lucernae** [comment]
@boney-bun can you only take changes from upstream for this method only:
`setup_geoserver`
`start_geoserver`

For the other, I'm still trying to merge it in another PR. I think travis failed because of these 2 method only.

File: pavement.py [hunk 1]
```diff
@@ -333,10 +398,18 @@ def start_django():
     """
     Start the GeoNode Django application
     """
-    bind = options.get('bind', '0.0.0.0:8000')
```
Comment (lucernae) [inline pavement.py]: Can you keep this one. It is needed to expose it to qgis server container.

File: pavement.py [hunk 2]
```diff
@@ -347,12 +420,7 @@ def start_geoserver(options):
     Start GeoServer with GeoNode extensions
     """
 
-    from geonode.settings import OGC_SERVER, INSTALLED_APPS
-
-    # only start if using Geoserver backend
-    if 'geonode.geoserver' not in INSTALLED_APPS:
-        return
-
```
Comment (lucernae) [inline pavement.py]: Keep this one, needed for QGIS Server, so it doesn't start Geoserver.

File: pavement.py [hunk 3]
```diff
@@ -436,31 +510,9 @@ def test(options):
...
-        sh("%s manage.py test %s.tests --noinput --liveserver=0.0.0.0:8000" % (
-            options.get('prefix'), '.tests '.join(GEONODE_APPS)))
-    except BuildFailure as e:
-        info('Tests failed! %s' % str(e))
-    else:
-        success = True
-    finally:
-        if 'geonode.geoserver' in INSTALLED_APPS:
-            # don't use call task here - it won't run since it already has
-            stop()
```
Comment (lucernae) [inline pavement.py]: Keep this one. It will be handled in my PR to merge upstream.

File: pavement.py [hunk 4]
```diff
@@ -436,31 +510,9 @@ def test(options):
...
+    sh("%s manage.py test %s.tests --noinput" % (options.get('prefix'),
+                                                 '.tests '.join(GEONODE_APPS)))
 
-    _reset()
```
Comment (lucernae) [inline pavement.py]: Keep this one, same reason as above.

File: pavement.py [hunk 5]
```diff
@@ -436,31 +510,9 @@ def test(options):
...
-    except BuildFailure as e:
-        info('Tests failed! %s' % str(e))
-    else:
-        success = True
-    finally:
-        if 'geonode.geoserver' in INSTALLED_APPS:
-            # don't use call task here - it won't run since it already has
-            stop()
+    sh("%s manage.py test %s.tests --noinput" % (options.get('prefix'),
+                                                 '.tests '.join(GEONODE_APPS)))
```
Comment (lucernae) [inline pavement.py]: Delete this.

File: pavement.py [hunk 6]
```diff
@@ -491,8 +543,8 @@ def test_integration(options):
             sh('sleep 30')
             call_task('setup_data')
         sh(('python manage.py test %s'
-           ' --noinput -v 3 --liveserver=0.0.0.0:8000' % name))
-    except BuildFailure, e:
+            ' --noinput --liveserver=localhost:8000' % name))
```
Comment (lucernae) [inline pavement.py]: change localhost to 0.0.0.0 because we need to expose it to QGIS Server containers

File: pavement.py [hunk 7]
```diff
@@ -749,4 +803,4 @@ def justcopy(origin, target):
     elif os.path.isfile(origin):
         if not os.path.exists(target):
             os.makedirs(target)
-        shutil.copy(origin, target)
+        shutil.copy(origin, target)
```
Comment (lucernae) [inline pavement.py]: Add newline

File: pavement.py [hunk 8]
```diff
@@ -117,13 +129,61 @@ def _install_data_dir():
...
+            m = re.search('logoutUri>([^<]+)', xml)
+            xml = xml[:m.start(1)] + "http://localhost:8000/account/logout/" + xml[m.end(1):]
+            with open(config, 'w') as f:
+                f.write(xml)
+    except Exception as e:
+        print(e)
+
+    try:
+        config = path(
+            'geoserver/data/security/role/geonode REST role service/config.xml')
```
Comment (lucernae) [inline pavement.py]: This path is weird. Was it coming from upstream?

Comment (boney-bun) [reply pavement.py]: yes, it's so weird.
indeed, it is from upstream:
https://github.com/GeoNode/geonode/blob/master/pavement.py#L178

**Comment by lucernae** [comment]
Thank's @boney-bun. LGTM. You still haven't fixed one issue. Please Squash merge after you resolved that one.

File: pavement.py [hunk 9]
```diff
@@ -749,4 +831,5 @@ def justcopy(origin, target):
     elif os.path.isfile(origin):
         if not os.path.exists(target):
             os.makedirs(target)
-        shutil.copy(origin, target)
+
+        shutil.copy(origin, target)
```
Comment (Gustry) [inline pavement.py]: @lucernae was talking about one new line after that one, not before ;-) Always one empty line at the end of a python file

Comment (boney-bun) [reply pavement.py]: oops...
thanks for the feedback @Gustry
wondering why it could pass the travis before.

Comment (Gustry) [reply pavement.py]: I don't think travis is checking code layout on this `pavement.py`, it's only a deployment file. I would say only on the django app itself. (I don't know, I've never checked the travis config)

Comment (lucernae) [reply pavement.py]: Yes, travis didn't do pep8 check for geonode. But we should still do the best practice :+1: 
GeoNode upstream also gradually try to follow pep8 guideline but haven't put the test yet on travis.

Comment (boney-bun) [reply pavement.py]: nice explanations.
thanks @lucernae @Gustry

**Comment by boney-bun** [comment]
thanks @lucernae 
i'll merge this.

**Pull request merged**